core/main/src/broker/thunder_broker.rs

```diff
@@ -0,0 +1,222 @@
+        let method = collection.pop();
+        let callsign = collection.join(".");
+        (callsign, method)
+    }
+}
+
+impl EndpointBroker for ThunderBroker {
+    fn get_broker(endpoint: RuleEndpoint, callback: BrokerCallback) -> Self {
+        Self::start(endpoint, callback)
+    }
+
+    fn get_sender(&self) -> BrokerSender {
+        self.sender.clone()
+    }
+
+    fn prepare_request(
+        &self,
+        rpc_request: &super::endpoint_broker::BrokerRequest,
+    ) -> Result<Vec<String>, RippleError> {
+        let mut requests = Vec::new();
+        let rpc = rpc_request.clone().rpc;
+        let id = rpc.ctx.call_id;
+        let app_id = rpc.ctx.app_id;
+        let (callsign, method) = Self::get_callsign_and_method_from_alias(&rpc_request.rule.alias);
+        if method.is_none() {
+            return Err(RippleError::InvalidInput);
+        }
+        let method = method.unwrap();
+        if rpc_request.rpc.is_subscription() {
+            let listen = rpc_request.rpc.is_listening();
+            let notif_id = {
+                let mut sub_map = self.subscription_map.write().unwrap();
+
+                if listen {
+                    if let Some(cleanup) = sub_map.insert(app_id, rpc_request.clone()) {
+                        requests.push(
+                            json!({
+                                "jsonrpc": "2.0",
+                                "id": cleanup.rpc.ctx.call_id,
+                                "method": format!("{}.{}", callsign, "unregister".to_owned()),
+                                "params": {
+                                    "event": method,
+                                    "id": format!("{}", cleanup.rpc.ctx.call_id)
+                                }
+                            })
+                            .to_string(),
+                        )
+                    }
+                    id
+                } else if let Some(v) = sub_map.remove(&app_id) {
+                    v.rpc.ctx.call_id
+                } else {
+                    id
+                }
+            };
+            requests.push(
+                json!({
+                    "jsonrpc": "2.0",
+                    "id": id,
+                    "method": format!("{}.{}", callsign, match listen {
+                        true => "register",
+                        false => "unregister"
+                    }),
+                    "params": json!({
+                        "event": method,
+                        "id": format!("{}", notif_id)
+                    })
+                })
+                .to_string(),
+            )
+        } else {
+            let request = Self::update_request(rpc_request)?;
+            requests.push(request)
+        }
+
+        Ok(requests)
+    }
+
+    /// Default handler method for the broker to remove the context and send it back to the
+    /// client for consumption
+    fn handle_response(result: &[u8], callback: BrokerCallback) {
```

Comment:
There's a duplicate of this in `endpoint_broker.rs`, consider making common.

Reply:
Not exactly the same, Thunder broker does `let updated_data = Self::update_response(&data);` on top of the handle_response. Which is necessary to convert event messages  which have different JSON Rpc structure between Thunder and Firebolt